Multiverse-io/companies_service
# Add RabbitMQ integration with proof of concept publisher and consumer

### What does this PR do?
This PR is adding the RabbitMQ integration with companies service.
As a proof of concept we have a dummy publisher and consumer.

### Context and implementation details

#### ID field
Something I noticed is that we use the ID field in the event payload to setup partitions in rabbitmq.
I'm not 100% sure why this is the case but I added a comment so we keep in mind that some ID is always mandatory in the payload.

#### Exchanges/Channels
I set up just one exchange (`companies_service.topic_exchange`) and one channel (`main`) since after doing some research I couldn't find any reason to split things further than that. I believe we can safely stick to just this exchange/channel for all messages published by the companies service to avoid complicating this more than needed.

But when it comes to consuming messages we will need to bind to other exchanges since thats where the messages are being published to :T

#### Queue names
Also since I believe almost always a queue will bind to a single specific routing key I used the `routing_key.queue` naming convention for the queue names, that way its immediately obvious which kind of messages are being sent into this queue.

#### Demo/Staging
I'm not sure how this is gonna work, we will see when this PR gets merged and deployed.

Either it could require some configuration to make it work or maybe it will work out of the box with the service laucher default configs. 🤔 

### How to test
1. Run `mix setup` to set the application up and create the RabbitMQ exchange locally.
2. Run `iex -S mix phx.server` to run the server in interactive mode
3. Run something like `CompaniesService.Publishers.Dummy.publish(%{id: 1, name: "foo"})` in the iex console
4. Check if both the event publication log and the event consuming log are displayed


### Links
[Ticket](https://app.shortcut.com/multiverse/story/45579/add-rabbitmq-setup-to-companies-service)
